File: thennow/modules/thennow-pose/ios/ThenNowPoseModule.swift
import ExpoModulesCore
import MLKitPoseDetectionAccurate
import MLKitPoseDetectionCommon
import MLKitVision

private let landmarkNames: [PoseLandmarkType: String] = [
  .nose: "nose",
  .leftEye: "leftEye",
  .rightEye: "rightEye",
  .leftEar: "leftEar",
  .rightEar: "rightEar",
  .leftShoulder: "leftShoulder",
  .rightShoulder: "rightShoulder",
  .leftElbow: "leftElbow",
  .rightElbow: "rightElbow",
  .leftWrist: "leftWrist",
  .rightWrist: "rightWrist",
  .leftHip: "leftHip",
  .rightHip: "rightHip",
  .leftKnee: "leftKnee",
  .rightKnee: "rightKnee",
  .leftAnkle: "leftAnkle",
  .rightAnkle: "rightAnkle",
]

public class ThenNowPoseModule: Module {
  public func definition() -> ModuleDefinition {
    Name("ThenNowPose")

    AsyncFunction("detectPose") { (uri: String, promise: Promise) in
      let path = uri.hasPrefix("file://") ? String(uri.dropFirst(7)) : uri
      guard let image = UIImage(contentsOfFile: path.removingPercentEncoding ?? path) else {
        promise.reject("E_IMAGE_LOAD", "Could not load image at \(uri)")
        return
      }

      let options = AccuratePoseDetectorOptions()
      options.detectorMode = .singleImage
      let detector = PoseDetector.poseDetector(options: options)

      let visionImage = VisionImage(image: image)
      visionImage.orientation = image.imageOrientation

      detector.process(visionImage) { poses, error in
        if let error = error {
          promise.reject("E_POSE_DETECT", "Pose detection failed: \(error.localizedDescription)")
          return
        }
        guard let pose = poses?.first else {
          promise.resolve(nil)
          return
        }
        var landmarks: [String: [String: Double]] = [:]
        for landmark in pose.landmarks {
          guard let name = landmarkNames[landmark.type] else { continue }
          landmarks[name] = [
            "x": Double(landmark.position.x),
            "y": Double(landmark.position.y),
            "likelihood": Double(landmark.inFrameLikelihood),
          ]
        }
        promise.resolve([
          "width": Double(image.size.width * image.scale),
          "height": Double(image.size.height * image.scale),
          "landmarks": landmarks,
        ] as [String: Any])
      }
    }
  }
}

File: thennow/modules/thennow-pose/android/src/main/java/expo/modules/thennowpose/ThenNowPoseModule.kt
package expo.modules.thennowpose

import android.net.Uri
import com.google.mlkit.vision.common.InputImage
import com.google.mlkit.vision.pose.PoseDetection
import com.google.mlkit.vision.pose.PoseLandmark
import com.google.mlkit.vision.pose.accurate.AccuratePoseDetectorOptions
import expo.modules.kotlin.Promise
import expo.modules.kotlin.exception.CodedException
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

private val LANDMARK_NAMES = mapOf(
  PoseLandmark.NOSE to "nose",
  PoseLandmark.LEFT_EYE to "leftEye",
  PoseLandmark.RIGHT_EYE to "rightEye",
  PoseLandmark.LEFT_EAR to "leftEar",
  PoseLandmark.RIGHT_EAR to "rightEar",
  PoseLandmark.LEFT_SHOULDER to "leftShoulder",
  PoseLandmark.RIGHT_SHOULDER to "rightShoulder",
  PoseLandmark.LEFT_ELBOW to "leftElbow",
  PoseLandmark.RIGHT_ELBOW to "rightElbow",
  PoseLandmark.LEFT_WRIST to "leftWrist",
  PoseLandmark.RIGHT_WRIST to "rightWrist",
  PoseLandmark.LEFT_HIP to "leftHip",
  PoseLandmark.RIGHT_HIP to "rightHip",
  PoseLandmark.LEFT_KNEE to "leftKnee",
  PoseLandmark.RIGHT_KNEE to "rightKnee",
  PoseLandmark.LEFT_ANKLE to "leftAnkle",
  PoseLandmark.RIGHT_ANKLE to "rightAnkle",
)

class ThenNowPoseModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("ThenNowPose")

    AsyncFunction("detectPose") { uri: String, promise: Promise ->
      val context = appContext.reactContext
        ?: return@AsyncFunction promise.reject(CodedException("E_NO_CONTEXT", "React context unavailable", null))

      val image: InputImage
      try {
        image = InputImage.fromFilePath(context, Uri.parse(uri))
      } catch (e: Exception) {
        promise.reject(CodedException("E_IMAGE_LOAD", "Could not load image: ${e.message}", e))
        return@AsyncFunction
      }

      val options = AccuratePoseDetectorOptions.Builder()
        .setDetectorMode(AccuratePoseDetectorOptions.SINGLE_IMAGE_MODE)
        .build()
      val detector = PoseDetection.getClient(options)

      detector.process(image)
        .addOnSuccessListener { pose ->
          val landmarks = mutableMapOf<String, Map<String, Any>>()
          for (lm in pose.allPoseLandmarks) {
            val name = LANDMARK_NAMES[lm.landmarkType] ?: continue
            landmarks[name] = mapOf(
              "x" to lm.position.x.toDouble(),
              "y" to lm.position.y.toDouble(),
              "likelihood" to lm.inFrameLikelihood.toDouble(),
            )
          }
          if (landmarks.isEmpty()) {
            promise.resolve(null)
          } else {
            promise.resolve(
              mapOf(
                "width" to image.width,
                "height" to image.height,
                "landmarks" to landmarks,
              )
            )
          }
        }
        .addOnFailureListener { e ->
          promise.reject(CodedException("E_POSE_DETECT", "Pose detection failed: ${e.message}", e))
        }
        .addOnCompleteListener {
          detector.close()
        }
    }
  }
}
